### DISCUSSION TITLE
Subimage Search magick++ api

### DISCUSSION THREAD

**Discussion by sdikshit786** [ImageMagick/ImageMagick#5944] [Help] [answered]
### ImageMagick version

7.1.0-0

### Operating system, version and so on

Linux(Ubuntu18.04)

### Description

I am trying to use the subimage search C++ api inside the image class. I have following queries please guide:

1. For subimage  search to work is it mandatory to enable MAGICKCORE_HDRI_SUPPORT and MAGICKCORE_FFTW_DELEGATE.

2. For most of the times I get the images are too disimilar error.
3. If the subimage occurs more than once how to detect that.
4. What is the role of min_similarity threshold and how to decide that?

**Comment by fmw42** [comment, marked as answer]
1) Only for NCC metric. (The NCC metric algorithm was updated for about 100x speedup, but requires HDRI and FFTW)

2) use -disimilarity-threshold 1 (or higher depending upon the metric) to keep the process from stopping if the images are too dissimilar.

3) Use a mask to zero out the region around the first peak, then look for the next peak. Use 

`magick identify -precision 5 -define identify:locate=maximum -define identify:limit=1 your_image.suffix`

Also see unix bash script at http://www.fmwconcepts.com/imagemagick/maxima/index.php as an example of how to do that

4) I am not familiar with that feature, but I suspect it throws out low value so you do not get too many bad matches. Each metric has a different best match value. Some like NCC are at 1 and some like RMSE are at 0.  Both vary between 0 and 1. So pick a value say around 0.9 for NCC and 0.1 for RMSE, if you really want to do that. I have never used it and just leave it out, ie. at its default which is likely all matches.

**Comment by fmw42** [comment]
There is no min_similarity threshold. Do you mean -similarity-threshold? If so, then from the documentation

> If this setting is used, then the search will stop as soon as it finds a match whose metric is less than or equal to the value. A partially filled second output image will result. Using a value of zero, will cause the search to stop at the first perfect match it finds. If this setting is left off, then the search will proceed to completion or as limited by -dissimilarity-threshold.

See https://imagemagick.org/script/command-line-options.php#similarity-threshold

**Comment by sdikshit786** [comment]
Thanks @fmw42 for the support
